File: point_cloud/lidar.py
import numpy as np
import itertools, pdb

# TODO Add documentation about the fact that these are all unit vectors
# TODO Test out rotating and plotting that as well
class Lidar(object):
    """LIDAR object
    
    Hold the pointing direction of the camera in body frame

    Provide methods to derive all the unit vectors for the LIDAR measurements

    Another method to transform into another frame given a transformation matrix

    """

    def __init__(self, view_axis=np.array([1, 0, 0]),
                 up_axis=np.array([0, 0, 1]),
                 fov=np.deg2rad(np.array([7, 7])), sigma=0.2,
                 dist=1, num_step=3):
        """Initialize the object

        cam_axis : define the pointing direction of center of LIDAR in the
        body fixed frame

        fov : horizontal and vertical field of view in radians

        sigma : 3 sigma uncertainty in depth measurements
        """
        

        right_axis = np.cross(view_axis, up_axis)

        H = np.tan(fov[1]/2) * dist
        W = np.tan(fov[0]/2) * dist
        
        # center of view fustrum
        c = view_axis * dist
        
        # corners of the fustrum
        hsteps = np.linspace(-H, H, num_step)
        wsteps = np.linspace(-W, W, num_step)
        
        self.lidar_arr = np.zeros((num_step**2, 3))
        # define all the unit vectors for the sensor
        for ii, (h, w) in enumerate(itertools.product(hsteps, wsteps)):
            lidar_vec = c + h * up_axis + w * right_axis
            self.lidar_arr[ii, :] = lidar_vec / np.linalg.norm(lidar_vec)

        # save things to object
        self.up_axis = up_axis
        self.view_axis = view_axis
        self.right_axis = right_axis

    def rotate_fov(self, R_body2frame):
        """Rotate the entire FOV by a given rotation matrix
        """
        # initialize an object

        # rotate all the vector by the new rotation matrix
        self.lidar_arr = R_body2frame.dot(self.lidar_arr.T).T 

        return self.lidar_arr
    

File: src/lidar.cpp
#include "lidar.hpp"

#include <cmath>

#include <iostream>

Lidar::Lidar( void ) {
	mview_axis << 1, 0, 0;
	mup_axis << 0, 0, 1;
	mfov << 0.12217, 0.12217;
	msigma = 0.2;
	mdist = 1;
	mnum_steps = 3;

	this->init();
}


void Lidar::init( void ) {
	mright_axis = mview_axis.cross(mup_axis);

	double H, W;
	H = std::tan(mfov(1)/2) * mdist;
	W = std::tan(mfov(0)/2) * mdist;
    
	// center of the view fustrum
	Eigen::Vector3d c;
	c = mview_axis * mdist;

	// corners of the fustrum
	Eigen::Matrix<double, Eigen::Dynamic, 1> hsteps(mnum_steps), wsteps(mnum_steps);
	hsteps = Eigen::VectorXd::LinSpaced(mnum_steps, -H, H);
	wsteps = Eigen::VectorXd::LinSpaced(mnum_steps, -W, W);

	// define  all the unit vectors for the sensor
	mlidar_array.resize(pow(mnum_steps, 2), 3);
	Eigen::RowVector3d lidar_vec;
	for (int ii = 0; ii < hsteps.size(); ++ii) {
		for (int jj = 0; jj < wsteps.size(); ++jj) {
			lidar_vec = c + hsteps(ii) * mup_axis + wsteps(jj) * mright_axis;
			mlidar_array.row(ii *  mnum_steps + jj) << lidar_vec / lidar_vec.norm();
		}
	}

}

Eigen::Matrix<double, Eigen::Dynamic, 3> Lidar::rotate_fov(const Eigen::Ref<const Eigen::Matrix<double, 3, 3> > &R_body2frame) {
    Eigen::Matrix<double, Eigen::Dynamic, 3> lidar_arr(mlidar_array.rows(), 3);
    lidar_arr = (R_body2frame * mlidar_array.transpose()).transpose();
	return lidar_arr;
}

Eigen::Matrix<double, Eigen::Dynamic, 3> Lidar::define_targets(const Eigen::Ref<const Eigen::RowVector3d> &pos,
                                                               const Eigen::Ref<const Eigen::Matrix<double, 3, 3> > &R_b2f,
                                                               const double &dist) {

    Eigen::Matrix<double, Eigen::Dynamic, 3> targets(mnum_steps * mnum_steps, 3);
    targets = (dist * rotate_fov(R_b2f)).rowwise() + pos;
    return targets;
}

Eigen::RowVector3d Lidar::define_target(const Eigen::Ref<const Eigen::RowVector3d> &pos,
                                       const Eigen::Ref<const Eigen::Matrix<double, 3, 3> > &R_b2f,
                                       const double &dist) {
    
    Eigen::RowVector3d target;
    target = (R_b2f * (dist * mview_axis)).transpose() + pos;
    return target;
}

Eigen::Vector3d Lidar::get_view_axis() {
    return mview_axis;
}

Eigen::Matrix<double, Eigen::Dynamic, 3> Lidar::get_lidar_array() {
    return mlidar_array;
}

Eigen::Vector3d Lidar::get_up_axis() {
    return mup_axis;
}
